File: CSharpCodeBase/components/controlsystem/actions/actioninteract.lua
local Action = require 'components.controlsystem.action'
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'

local STATE_MOVING = 0
local STATE_REQUEST_INTERACT = 1
local STATE_INTERACT = 2

local ActionInteract = Class({__includes=Action, init = function (self, entity, target)
      Action.init(self, 1, false, 'interact')
      self.entity = entity
      self.target = target
      self.state = STATE_MOVING
end})

function ActionInteract:OnMoveComplete()
  self.state = STATE_REQUEST_INTERACT
  self.entity.combat:Attack(self.target, self.entity.weaponContainer.interactWeapon)
end

function ActionInteract:OnStartRunning()
  self.state = STATE_MOVING
  self.entity.mover:SetGoal(self.target, nil, self.entity.weaponContainer:GetAttackDistance(self.entity.weaponContainer.interactWeapon))
end

function ActionInteract:OnStopRunning()
  if self.state == STATE_MOVING then
    self.entity.mover:Stop()
  end
end

function ActionInteract:Update()
  if self.state == STATE_MOVING and not self.target.interactable then
    return true 
  end
  if self.state == STATE_MOVING and not self.entity.mover:IsHaveGoal() then
    self:OnMoveComplete()
  end
  
  return self.state == STATE_INTERACT and not self.entity.mecanim:CheckStateName("Action")
end

function ActionInteract:OnEvent(data)
  if data == "punch" then            
    self.state = STATE_INTERACT
    if self.target.interactable and self.entity.weaponContainer:CanAttack(self.target)then
      self.target:Message("OnInteract", self.entity)
    end    
  end
end

function ActionInteract:GetPriority()
  return self.state == STATE_MOVING and 1 or 3
end

function ActionInteract:IsContinuous()
  return self.state == STATE_MOVING
end

return ActionInteract

File: CSharpCodeBase/components/controlsystem/actions/actioninteract.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class actioninteract :action{
 var STATE_MOVING = 0;
 var STATE_REQUEST_INTERACT = 1;
 var STATE_INTERACT = 2;
 public void init (self, entity, target);
       Action.init(self, 1, false, "interact");
       this.entity = entity;
       this.target = target;
       this.state = STATE_MOVING;
 }})
 public void OnMoveComplete(){
   this.state = STATE_REQUEST_INTERACT;
   this.entity.combat:Attack(this.target, this.entity.weaponContainer.interactWeapon);
 }
 public void OnStartRunning(){
   this.state = STATE_MOVING;
   this.entity.mover:SetGoal(this.target, null, this.entity.weaponContainer:GetAttackDistance(this.entity.weaponContainer.interactWeapon));
 }
 public void OnStopRunning(){
   if(this.state == STATE_MOVING  ){ 
     this.entity.mover:Stop();
   }
 }
 public void Update(){
   if(this.state == STATE_MOVING  &&  not this.target.interactable  ){ 
     return true ;
   }
   if(this.state == STATE_MOVING  &&  not this.entity.mover:IsHaveGoal()  ){ 
     self:OnMoveComplete();
   }
   
   return this.state == STATE_INTERACT  &&  not this.entity.mecanim:CheckStateName("Action");
 }
 public void OnEvent(data){
   if(data == "punch"  ){             
     this.state = STATE_INTERACT;
     if(this.target.interactable  &&  this.entity.weaponContainer:CanAttack(this.target)then;
       this.target:Message("OnInteract", this.entity);
     }    
   }
 }
 public void GetPriority(){
   return this.state == STATE_MOVING  &&  1  ||  3;
 }
 public void IsContinuous(){
   return this.state == STATE_MOVING;
 }
 }}